### PR TITLE
PR2: Checkout contract

### PR DESCRIPTION
**Pull request by harshasomisetty**
All checkout contract code

### CODE REVIEW

File: programs/checkout/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+        Ok(())
+    }
+
+    pub fn cancel_bid(ctx: Context<CancelBid>) -> Result<()> {
+        let bidder = &ctx.accounts.bidder;
+
+        let mut heap = ctx.accounts.bid_orders.load_mut()?;
+
+        // Need a clever way to somehow know the bid price after the let mut heap declaration
+        let bid_price_sol = heap.heap.cancel_bid(bidder.key());
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs]: We discussed this, but this function needs to be adjusted to isolate an exact order.

File: programs/checkout/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+        **ctx
+            .accounts
+            .escrow_sol
+            .to_account_info()
+            .try_borrow_mut_lamports()? -= bid_price_sol;
+        **ctx.accounts.bidder.try_borrow_mut_lamports()? += bid_price_sol;
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs:86]: This seems dangerous. if `heap.cancel_bid` somehow returns a non-zero value for a mismatch, you can remove all SOL from the contract. I would mitigate this by hard failing in the case that you fail to cancel bid (just feels safer)

File: programs/checkout/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+        // msg!("canceling heap data: {}", heap.heap);
+
+        Ok(())
+    }
+
+    pub fn bid_floor(ctx: Context<BidFloor>) -> Result<()> {
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs]: This function name should probably be called `sell_to_floor`

File: programs/checkout/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+        ctx: Context<FulfillOrder>,
+        pubkey_to_remove: Pubkey,
+        checkout_auth_bump: u8,
+    ) -> Result<()> {
+        msg!("in remove_order pubkey: {}", &pubkey_to_remove.to_string());
+
+        let mut matched_orders = ctx.accounts.matched_orders.load_mut()?;
+        matched_orders
+            .trades
+            .remove_node_by_pubkey(pubkey_to_remove);
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs:148]: Assuming that this always just removes the first node that matches? I feel like better functionality would be to remove multiple. Not a huge issue

File: programs/checkout/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+    pub escrow_voucher: Account<'info, TokenAccount>,
+
+    #[account(
+        init,
+        payer = user,
+        space = std::mem::size_of::<SolWallet>() + 8,
+        seeds = [b"escrow_sol", exhibit.key().as_ref()],
+        bump
+    )]
+    pub escrow_sol: Account<'info, SolWallet>,
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs:220]: There's no reason to make this a struct. What you should do is seed it enough SOL for rent exemption `solana rent 0` and the call `system_instruction::transfer` to this escrow account as opposed to modifying the lamports directly

File: programs/checkout/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+    #[account(init,
+        payer = user,
+        space = std::mem::size_of::<BidOrders>() + 8,
+        seeds = [b"bid_orders", exhibit.key().as_ref()], bump
+    )]
+    pub bid_orders: AccountLoader<'info, BidOrders>,
+
+    #[account(init, payer = user, space = 8+std::mem::size_of::<CheckoutAuth>(), seeds=[b"checkout_auth", exhibit.key().as_ref()], bump)]
+    pub checkout_auth: Account<'info, CheckoutAuth>,
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs:198]: Probably doesn't need to have its own struct

File: programs/checkout/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+pub mod checkout {
+
+    use solana_program::native_token::LAMPORTS_PER_SOL;
+
+    use super::*;
+
+    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
+        msg!("In initializer");
+
+        let list = LinkedList::initialize();
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs:29]: This allocates on the stack. You could potentially run out of memory this way if you want to increase the DLL size

File: programs/checkout/src/lib.rs (hunk 2 above)
Comment (jarry-xiao) [reply programs/checkout/src/lib.rs:86]: See the comment below about using `system_instruction::transfer`

File: programs/checkout/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+};
+
+use solana_program;
+use solana_program::{account_info::AccountInfo, program::invoke, system_instruction};
+
+use exhibition::program::Exhibition;
+use exhibition::{self, Exhibit};
+
+pub mod state;
+use state::heap::Heap;
```
Comment (jarry-xiao) [inline programs/checkout/src/lib.rs:14]: I'd probably rename this structs to what the represent (Orderbook, CheckoutQueue). The actual data structures should ideally be imported

File: programs/checkout/src/lib.rs (hunk 2 above)
Comment (fabioibanez) [reply programs/checkout/src/lib.rs:86]: heap.cancel_bid hard fails, still need to fix second comment on `system_instruction::transfer`.

File: programs/checkout/src/lib.rs (hunk 6 above)
Comment (harshasomisetty) [reply programs/checkout/src/lib.rs:198]: fixed in future iterations

File: programs/checkout/src/lib.rs (hunk 2 above)
Comment (harshasomisetty) [reply programs/checkout/src/lib.rs:86]: fixed transfer function

**Pull request closed without merging**